android: detect native ABI and API level correctly

lunagl opened the pull request:
Fixes ABI detection and adds API level detection (#24630).

I was planning on also including a fix for #24860, but I am not sure how to proceed due to the `faccessat` syscall not supporting flags (#16606).

lunagl committed:
android: detect native ABI and API level correctly

ABI detection previously did not take into account the non-standard
directory structure of Android. This has been fixed.

The API level is detected by running `getprop ro.build.version.sdk`,
since we don't want to depend on bionic, and reading system properties
ourselves is not trivially possible.

lunagl commented on lib/std/zig/system.zig:
```diff
@@ -449,6 +449,17 @@ pub fn resolveTargetQuery(query: Target.Query) DetectError!Target {
         }
     }
 
+    if (result.abi.isAndroid() and query.os_tag == null and query.android_api_level == null) {
```
At the moment the documentation for `Query.android_api_level` states that `os_tag == null` means native. But it might make more sense to switch that to `query.abi == null`, what do you think?

alexrp commented on lib/std/zig/system.zig:
```diff
@@ -1294,6 +1310,39 @@ fn elfInt(is_64: bool, need_bswap: bool, int_32: anytype, int_64: anytype) @Type
     }
 }
 
+fn detectAndroidApiLevel() !u32 {
```
Returning errors would probably be good enough in this function; I don't think we need the logging.

alexrp commented:
Would be good to know what kind of Android system(s) this has been tested on.

alexrp commented on lib/std/zig/system.zig:
```diff
@@ -1294,6 +1310,31 @@ fn elfInt(is_64: bool, need_bswap: bool, int_32: anytype, int_64: anytype) @Type
+    const api_level = try std.fmt.parseInt(u32, stdout_buf[0 .. n - 1], 10);
+
+    const term = try child.wait();
+    if (term != .Exited or term.Exited != 0) return error.TerminatedAbnormally;
```
```suggestion
    if (term != .Exited or term.Exited != 0) return error.GetpropTerminatedAbnormally;
```

Would be fine to be a bit more specific here.

alexrp commented on lib/std/zig/system.zig:
```diff
@@ -1294,6 +1310,31 @@ fn elfInt(is_64: bool, need_bswap: bool, int_32: anytype, int_64: anytype) @Type
+    // Currently API levels are two-digit numbers, but we want to make sure we never read a partial value.
+    var stdout_buf: [92 + 1]u8 = undefined;
+    var reader = child.stdout.?.readerStreaming(&.{});
+    const n = try reader.interface.readSliceShort(&stdout_buf);
+    const api_level = try std.fmt.parseInt(u32, stdout_buf[0 .. n - 1], 10);
```
It would be reasonable to catch any errors here and return something like `error.GetpropOutputMalformed`.

lunagl commented:
I have tested this on a Google Pixel 7 running Android 16 (API 36). Since it only relies on `/system/bin/getprop` and the `ro.build.version.sdk` property, it should be pretty universally supported.

alexrp commented on lib/std/zig/system.zig:
```diff
@@ -1294,6 +1310,31 @@ fn elfInt(is_64: bool, need_bswap: bool, int_32: anytype, int_64: anytype) @Type
+    child.stdout_behavior = .Pipe;
+    child.stderr_behavior = .Ignore;
+
+    try child.spawn();
```
Hmm. This (and the `try child.wait()` call below) will add a bunch of potential errors that [would need to](https://github.com/ziglang/zig/pull/24868/files#r2280453774) become part of `DetectError`. I'm not convinced that that's actually particularly helpful to a caller of `resolveTargetQuery`; I suggest marking `unreachable` those errors that there's no reason to expect would ever happen in this specific scenario, and mapping the rest to 1-2 higher level errors as appropriate that mention `Getprop` in the error name(s).

lunagl commented on lib/std/zig/system.zig:
```diff
@@ -449,6 +449,17 @@ pub fn resolveTargetQuery(query: Target.Query) DetectError!Target {
+        if (detectAndroidApiLevel()) |api_level| {
+            result.os.version_range.linux.android = api_level;
+        } else |e| {
+            std.log.warn(
+                "Could not determine native Android API level ({s}), falling back to default",
+                .{@errorName(e)},
+            );
```
One thing that could happen with the current setup is that a user on a Linux host sets `query.abi` to Android, but leaves `query.os_tag` at `null` (I think, if there's no other logic to prevent that somewhere), making us believe we're building for the native system when we really aren't.

And generally I assumed that knowing the exact native API level isn't super important, so it might make sense to only warn and still let compilation succeed, but I'm not sure.

alexrp replied:
> One thing that could happen with the current setup is that a user on a Linux host sets `query.abi` to Android, but leaves `query.os_tag` at `null` (I think, if there's no other logic to prevent that somewhere), making us believe we're building for the native system when we really aren't.

If you compile for `x86_64-native-android`, you actually effectively get `x86_64-native-native`. I'm... not really sure what to make of that. It does at least indicate that this is a poorly defined / poorly tested state. Need to think a bit on that.

> And generally I assumed that knowing the exact native API level isn't super important, so it might make sense to only warn and still let compilation succeed, but I'm not sure.

Well, the fact that it affects something as fundamental as TLS codegen indicates that it's a fair bit more important than e.g. the libc version on other targets. In a sense, the API level is of similar importance to the OS version on Android.

alexrp commented on lib/std/zig/system.zig:
```diff
@@ -1294,6 +1318,31 @@ fn elfInt(is_64: bool, need_bswap: bool, int_32: anytype, int_64: anytype) @Type
 }
 
+fn detectAndroidApiLevel() !u32 {
+    var alloc_buf: [384]u8 = undefined;
```
Why this buffer size in particular?

lunagl replied:
Technically from what I can tell only argv plus it's elements and envp (which is empty) need to be allocated, but because `spawn` uses an ArenaAllocator it was difficult to determine the exact memory required (and I don't know if it's 100% deterministic, I believe 256 worked once and then failed on another try), so I started with 128, tried 256, and then 384 (256 + 128) as a value that didn't OOM. It's not super pretty I agree, but since we don't have a proper heap allocator here I'm not sure what else to do

rohlem replied:
I don't have a good strategy for this either. If this remains a heuristic guessed value, then a same-line comment that documents this fact for the future would be helpful.

lunagl replied:
I don't think my understanding of the codebase is deep enough to make that call, I could certainly implement it tho. Right now it would only be used for the API level detection, I'm not sure if other parts of `resolveTargetQuery` would benefit from an allocator as well.

alexrp replied:
If you want to take a stab at it, please go ahead, though I suggest doing it as a separate commit within the PR.

alexrp commented on lib/std/zig/system.zig:
```diff
@@ -1294,6 +1318,31 @@ fn elfInt(is_64: bool, need_bswap: bool, int_32: anytype, int_64: anytype) @Type
+    var stdout_buf: [92 + 1]u8 = undefined;
+    var reader = child.stdout.?.readerStreaming(&.{});
+    const n = try reader.interface.readSliceShort(&stdout_buf);
+    const api_level = std.fmt.parseInt(u32, stdout_buf[0 .. n - 1], 10) catch return error.GetpropOutputMalformed;
+
+    const term = try child.wait();
+    if (term != .Exited or term.Exited != 0) return error.GetpropTerminatedAbnormally;
```
To clarify, I suggested returning these specific errors because they seemed reasonable to bubble up as part of `DetectError`. But I think maybe you're right that that's leaking implementation details.

So, on second thought, let's log these error conditions as warnings as you did before. They indicate that an important contract of `getprop` and/or `ro.build.version.sdk` has been broken, and we'll definitely want to know about that. We should still ultimately return `ApiLevelQueryFailed` to the caller of `resolveTargetQuery` though.

alexrp replied:
Well, the `errdefer` path could exclude the errors that have more specific logging.

lunagl commented on lib/std/zig/system.zig:
```diff
@@ -449,6 +451,23 @@ pub fn resolveTargetQuery(query: Target.Query) DetectError!Target {
         }
     }
 
+    if (result.abi.isAndroid() and query.os_tag == null and query.android_api_level == null) {
```
We definitely need to keep the check for `query.os_tag == null` to make sure we're building for the native target, other than that I'll add this.

lunagl replied:
I realized you were right and we do need the comptime `builtin.os.tag` check, since we don't want to compile the `detectAndroidApiLevel` function at all if we're not building for Linux (it broke the wasi build since wasi doesn't support child process spawning)

Vexu commented on lib/std/zig/system.zig:
```diff
@@ -1293,6 +1317,48 @@ fn elfInt(is_64: bool, need_bswap: bool, int_32: anytype, int_64: anytype) @Type
+fn detectAndroidApiLevel() !u32 {
+    comptime if (builtin.os.tag != .linux) unreachable;
+
+    errdefer |err| if (err != error.ApiLevelQueryFailed) {
```
Capturing the error in errdefer is going to be removed #23734

lunagl committed:
Pass allocator to zig.system.resolveTargetQuery

For now the allocator is only used in `detectAndroidApiLevel`, but there
are probably other places in the detection logic that could benefit from
this too.

lunagl commented:
I'm trying to figure out why the pipelines are failing here, the errors seem unrelated to my changes. Any ideas?

andrewrk commented:
This pull request is not ready for review because:
* It has conflicts that must be resolved via rebasing against latest origin/master.
* It is not passing the CI tests.

Since we have [moved development to Codeberg](https://ziglang.org/news/migrating-from-github-to-codeberg/), please open your pull request there if you would like to continue these efforts.

andrewrk commented:
https://codeberg.org/ziglang/zig/pulls/30056
